chapter/methodology.md

```diff
@@ -0,0 +1,964 @@
+Yet, during the linearization the location might be unstable or unknown for different items.
+Record fields for instance are processed in an arbitrary order rather than the order they are defined.
+Moreover, for nested records and record short notations, symbolic `Record` items are created which cannot be mapped to a physical location and are thus placed at the range `[0..=0]` in the beginning of the file.
+Maintaining constant insertion performance and item-referencing requires that the linearization is exclusively appended.
+Each of these cases, break the physical linearity of the linearization.
+
+NLS thus defers reordering of items.
+The language server uses a stable sorting algorithm to sort items by their associated span's starting position.
+This way, nesting of items with the same start location is preserved.
+Since several operations require efficient access to elements by `id`, which after the sorting does not correspond to the items index in the linearization, after sorting NLS creates an index mapping `id`s to the new actual indices.
+
+#### Resolving deferred access
+
+[Section @sec:variable-usage-and-static-record-access] introduced the `Deferred` type for `Usages`.
+Resolution of usages is deferred if chained destructors are used.
+This is especially important in recursive records where any value may refer to other fields of the record which could still be unresolved.
+
+As seen in [@fig:ncl-record-access], the items generated for each destructor only link to their parent item.
+Yet, the root access is connected to a known declaration.
+Since at this point all records are fully processed NLS is able to resolve destructors iteratively.
+
+First NLS collects all deferred usages in a queue.
+Each usage contains the *`id`* of the parent destructor as well as the *name* of the field itself represents.
+NLS then tries to resolve the base record for the usage by resolving the parent.
+If the value of the parent destructor is not yet known or a deferred usage, NLS will enqueue the destructor once again to be processed again later.
+In practical terms that is after the other fields of a common record.
+In any other case the parent consequently has to point to a record, either directly, through a record field or a variable.
+NLS will then get the `id` of the `RecordField` for the destructors *name* and mark the `Usage` as `Known`
+If no field with that name is present or the parent points to a `Structure` or `Unbound` usage, the destructor cannot be resolved in a meaningful way and will thus be marked `Unbound`.
+
+
+#### Resolving types
+
+<!-- TODO: link to background section -->
+As a necessity for type checking, Nickel generates type variables for any node of the AST which it hands down to the `Linearizer`.
+<!-- TODO: example for types? -->
+In order to provide meaningful information, the Language Server needs to derive concrete types from these variables.
+The required metadata needs to be provided by the type checker.
```

Comment:
```suggestion
<!-- TODO: link to background section -->
As a necessity for type checking, Nickel generates unification variables for any node of the AST which it hands down to the `Linearizer`.
In Nickel these variables act as type constraints which are solved later during the processing of a file.
As a result, when a node is encountered by NLS, its type is not fully resolved yet.
Similar to runtime processing, NLS needs to resolve the final types separately.
<!-- TODO: example for types? -->
```

Reply:
Suggestion, with some rewording and also insisting on type inference. Now that I think of it, type inference is the actual reason why you don't know the types right away: in a fully explicitly typed language, you wouldn't have to solve constraints to begin with.

```suggestion
<!-- TODO: link to background section -->
Nickel features type inference in order to relieve the programmer of the burden of writing a lot of redundant type annotations. In a typed block, the typechecker is able to guess the type of all the values, even when they are not explicitly annotated by the user. To do so, the typechecker generates constraints derived from inspecting the AST, and solve them along the way. As a consequence, when a node is first encountered by NLS, its type is not necessarily  known. There, the typechecker associate to the new node a so-called unification variable, which is a placeholder for a later resolved type. This unification variable is handed down to the `Linearizer`.

Similar to runtime processing, NLS needs to resolve the final types separately. After typechecking, NLS is eventually able to query the resolved type corresponding to an AST node using the associated unification variable.
<!-- TODO: example for types? -->
```

Reply:
I agree its clearer wrt. to the inference